### DISCUSSION TITLE
How to access `auth.admin.generateLink` from NextJS app router?

### DISCUSSION THREAD

**Discussion by sirvine** [supabase/supabase#15962] [Questions]
The docs and examples for NextJS auth helpers don't seem to address the situation where we want to retrieve a magic link and OTP directly and place those into our own email template. 

Specifically, `createServerComponentClient` doesn't allow me to invoke auth admin functions like `generateLink` because we're not using the service role key.

In the past, we'd use `createClient`. I tried that from within a NextJS app router app, and it kept returning the same magic link and OTP!

Has anyone tackled this kind of use case in Next 13.x using the app router?

**Comment by GaryAustin1** [comment]
You should be able to use createClient from supabase-js with service_role.  You don't use the auth-helpers clients as they set the Authorization header and would override service_role.   Set persistSession:false and detectSessionInUrl:false.

**Reply by sirvine** [reply]
@GaryAustin1 Thank you. I've tried this, and the client will submit a request for `generateLink` but the `email_otp` and `action_link` are always the same values for any given email address. 

Originally, I assumed this had to do with the client, but now I'm unsure.

**Reply by sirvine** [reply]
In case it's helpful, here is the relevant server action:

<img width="788" alt="CleanShot 2023-07-23 at 17 07 29@2x" src="https://github.com/supabase/supabase/assets/68167/df5104ed-cae4-4a50-9c35-63d368ef20e4">

And the response:

<img width="246" alt="CleanShot 2023-07-23 at 17 08 30@2x" src="https://github.com/supabase/supabase/assets/68167/e4626f65-627e-41ea-8338-47aaea365be5">
